lib/src/map/gestures/services/double_tap.dart

```diff
@@ -0,0 +1,52 @@
+part of 'base_services.dart';
+
+class DoubleTapGestureService extends DelayedGestureService {
+  DoubleTapGestureService({required super.controller});
+
+  /// A double tap gesture tap has been registered
+  void submit() {
```

Comment:
*Merging the inherited discussions here.*

I'd like to have them inherited too any did so once but reverted the change. The problem was that all `submit` functions create a different event type. We can however create a function to emit the event in every child class and let the parent handle the rest. 

Side note: If we rework use the event system to just have one camera move event that rules them all we don't need the additional function at all.

Reply:
> If we rework use the event system to just have one camera move event that rules them all we don't need the additional function at all.

How would this look?

> We can however create a function to emit the event in every child class and let the parent handle the rest.

This wouldn't however have the advantage of https://github.com/fleaflet/flutter_map/pull/1733#discussion_r1428962917, which I think is quite important.

Reply:
These are the the current method signatures (removed unused parameters):

- BaseGestureService
  - LongPressGestureService, SecondaryLongPressGestureService, TertiaryLongPressGestureService
    - `submit(LongPressStartDetails)`
  - ScrollWheelZoomGestureService
    - `submit(PointerScrollEvent)`
  - CtrlDragRotateGestureService
    - `start()`
    - `update(ScaleUpdateDetails)`
    - `end()`
  - DoubleTapDragZoomGestureService, DragGestureService, TwoFingerGesturesService
    - `start(ScaleStartDetails)`
    - `update(ScaleUpdateDetails)`
    - `end(ScaleEndDetails)`
  - DelayedGestureService
    - `setDetails(TapDownDetails)`
    - DoubleTapGestureService, TapGestureService, SecondaryTapGestureService, TertiaryTapGestureService
      - `submit()`

We can create a parent for these two:
- LongPressGestureService, SecondaryLongPressGestureService, TertiaryLongPressGestureService
- DoubleTapGestureService, TapGestureService, SecondaryTapGestureService, TertiaryTapGestureService

However there are unique gesture services like `ScrollWheelZoomGestureService` and `CtrlDragRotateGestureService` I don't know how we could use `@mustCallSuper`. Do you have an idea how to implement this so it gets enforced on every gesture service?

Reply:
Ok, I see. I'm not sure the current situation ideally expresses what inherits what.

There should be at least an intermediate `SingleShotGestureService` (which has `submit`) and `ProgressableGestureService` (which has `start`, `update`, `end`). We can use generic typing where necessary.  
This allows us to define base implementations of those methods which call `controller.stopAnimationRaw();` and enforce with `@mustCallSuper` & `@mustBeOverriden`, and the `base` class modifier to prevent implementations.  
`DelayedGestureService` might then be better as a mixin which provides `TapDownDetails`, or as a subclass of `SingleShotGestureService`, but I'm not sure about whether either of these would work, let alone which one is best.  
Have I missed something as to why this could not work? I probably have.

> However there are unique gesture services like `ScrollWheelZoomGestureService` and `CtrlDragRotateGestureService`

What makes these unique that a generic type wouldn't solve? And why doesn't `CtrlDragRotateGestureService` get args for `start` and `end`?

> `DoubleTapDragZoomGestureService`

Shouldn't this be a delayed gesture, as it relies on tap detection?

Reply:
The advantage of the current implementation is that we treat every gesture as individually as possible to have a simple while extendable class design.
We could use inheritance, generics or mixins. However what is the benefit that justifies the complexity, overhead less flexibility? What does this solve that we are currently not solving?

---

The name `DelayedGestureService` is a bit misleading. It is meant as a base service where the `TapDownDetails` get set first and gets submitted later on. The `DoubleTapDragZoomGestureService` uses a start(), update(), end() logic which doesn't get represented by the `DelayedGestureService`. I renamed it to `BaseDetailsGestureService`.

Reply:
Pull request that would resolve this thread in https://github.com/josxha/flutter_map/pull/2.

Reply:
pr merged